pkg/apiserver/apiserver.go

```diff
@@ -0,0 +1,196 @@
+	"k8s.io/apimachinery/pkg/version"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+const GroupName = "gitservice.devopsconsole.openshift.io"
+const GroupVersion = "v1alpha1"
+
+var (
+	Scheme             = runtime.NewScheme()
+	Codecs             = serializer.NewCodecFactory(Scheme)
+	SchemeBuilder      = runtime.NewSchemeBuilder(addKnownTypes)
+	AddToScheme        = SchemeBuilder.AddToScheme
+	SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: GroupVersion}
+)
+
+func addKnownTypes(scheme *runtime.Scheme) error {
+	scheme.AddKnownTypes(SchemeGroupVersion)
+	return nil
+}
+
+func init() {
+	utilruntime.Must(AddToScheme(Scheme))
+
+	// Setting VersionPriority is critical in the InstallAPIGroup call (done in New())
+	utilruntime.Must(Scheme.SetVersionPriority(SchemeGroupVersion))
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: GroupVersion})
+
+	unversioned := schema.GroupVersion{Group: "", Version: GroupVersion}
+	Scheme.AddUnversionedTypes(unversioned,
+		&metav1.Status{},
+		&metav1.APIVersions{},
+		&metav1.APIGroupList{},
+		&metav1.APIGroup{},
+		&metav1.APIResourceList{},
+	)
+}
+
+type ExtraConfig struct {
+	// Place you custom config here.
+}
+
+type Config struct {
+	GenericConfig *genericapiserver.RecommendedConfig
+	ExtraConfig   ExtraConfig
+}
+
+// GitAPIServer contains state for a Kubernetes cluster master/api server.
+type GitAPIServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	GenericConfig genericapiserver.CompletedConfig
+	ExtraConfig   *ExtraConfig
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
+}
+
+type GitSource struct {
+	Source string
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (cfg *Config) Complete() CompletedConfig {
+	c := completedConfig{
+		cfg.GenericConfig.Complete(),
+		&cfg.ExtraConfig,
+	}
+
+	c.GenericConfig.Version = &version.Info{
+		Major: "1",
+		Minor: "0",
+	}
+
+	return CompletedConfig{&c}
+}
+
+// New returns a new instance of GitAPIServer from the given config.
+func (c completedConfig) New() (*GitAPIServer, error) {
+	genericServer, err := c.GenericConfig.New("DevOps Console Git API Server", genericapiserver.NewEmptyDelegate())
+	if err != nil {
+		return nil, err
+	}
+
+	s := &GitAPIServer{
+		GenericAPIServer: genericServer,
+	}
+
+	apiGroupInfo := genericapiserver.NewDefaultAPIGroupInfo(GroupName, Scheme, metav1.ParameterCodec, Codecs)
+
+	if err := s.GenericAPIServer.InstallAPIGroup(&apiGroupInfo); err != nil {
+		return nil, err
+	}
+
+	installCompositionGitWebService(s)
+
+	return s, nil
+}
+
+func installCompositionGitWebService(gitServer *GitAPIServer) {
+	path := fmt.Sprintf("/apis/%s/%s/namespaces/{namespace}", GroupName, GroupVersion)
+	fmt.Println("WS PATH:" + path)
```

Comment:
can we remove println?

Reply:
Done in https://github.com/redhat-developer/git-service/pull/1/commits/b11eebe97c974ff98c10e4b8a7f97c3f5d82021f